run-tests.php

```diff
@@ -1545,8 +1545,14 @@ function run_all_tests_parallel($test_files, $env, $redir_tested) {
 								$sequentialTests = [];
 							}
 							// Batch multiple tests to reduce communication overhead.
+							// - When valgrind is used, communication overhead is relatively small,
+							//   so just use a batch size of 1.
+							// - If this is running a small enough number of tests,
+							//   reduce the batch size to give batches to more workers.
 							$files = [];
-							$batchSize = $shuffle ? 4 : 32;
+							$batchSize = $valgrind ? 1 : ($shuffle ? 4 : 32);
+							$averageFilesPerWorker = max(1, (int)ceil(count($files) / count($workerProcs)));
```

Comment:
Agree on the valgrind case, but the rest is less clear. For example, if you have 8 workers and 8*32 tests (and no conflicts), then it would make most sense to give each worker 32 each. But I think what's going to happen with your change is that the load will be distributed as

    worker 1: 32 (remaining: 224)
    worker 2: 28 (remaining: 196)
    worker 3: 25 (remaining: 171)
    worker 4: 22 (remaining: 149)
    ...

Basically, what you do here makes sense for the initial distribution if all workers are idle, but not when the workers are already busy.

Reply:
Yeah, that's a bug - I was looking at the wrong array and wanted the constant list of all tests. In fact, $files = [];

I should instead be using $TOTAL_FILE_COUNT = count($test_files); at the top of this function declaration, e.g. around the array_reverse